common/src/buffered-buffer.ts

```diff
@@ -0,0 +1,55 @@
+import { createAsyncQueue } from "./async-queue"
+
+const MAX_BUFFERED_BUFFER = 100;
+
+export class BufferedBuffer {
+    queue: ReturnType<typeof createAsyncQueue<Buffer | string>>
+
+    constructor() {
+        this.queue = createAsyncQueue<Buffer | string>();
+    }
+
+    append(data: Buffer | string) {
+        this.queue.submit(data);
+    }
+
+    close() {
+        this.queue.end();
+    }
+
+    onClose(fn: Function) {
+        this.queue.onEnd(fn);
+    }
+
+    /**
+     * The returned AsyncGenerator will concatenate together
+     * sequential Buffers, but leave strings unchanged.
+     */
+    async *generator(): AsyncGenerator<Buffer | string, void> {
+        while (!this.queue.ended) {
```

Comment:
I think what you want here is clear, followed by a concat.

Reply:
If clear results in empty set, then await dequeue

Reply:
Reasoning for doing it this way iteratively is to keep string messages intact but concat all buffers. A clear would then need an iteration through the list to find if there's a string, concat elements before the string, then requeue the rest of the list

Reply:
only the client side sends strings I think. it would probably be easier to follow by removing this this abstracted bufferedbuffer altogether and use the async queue directly on both sides.